src/prefect/runner/storage.py

```diff
@@ -262,22 +282,48 @@ async def pull_code(self) -> None:
             cmd = ["git"]
             # Add the git configuration, must be given after `git` and before the command
             cmd += self._git_config
-            # Add the pull command and parameters
-            cmd += ["pull", "origin"]
-            if self._branch:
-                cmd += [self._branch]
-            if self._include_submodules:
-                cmd += ["--recurse-submodules"]
-            cmd += ["--depth", "1"]
-            try:
-                await run_process(cmd, cwd=self.destination)
-                self._logger.debug("Successfully pulled latest changes")
-            except subprocess.CalledProcessError as exc:
-                self._logger.error(
-                    f"Failed to pull latest changes with exit code {exc}"
+
+            # If checking out a specific commit, fetch the latest changes and unshallow the repository if necessary
+            if self._commit_sha:
+                if await self.is_shallow_clone():
+                    cmd += ["fetch", "origin", "--unshallow"]
+                else:
+                    cmd += ["fetch", "origin", self._commit_sha]
```

Comment:
If a commit SHA is provided and the git repo is already present, chances are we don't need to perform another pull since individual commits are more immutable than branches. Could we check the current commit of the repo and only fetch if it doesn't match the provided commit SHA?